File: Bulwark.Core/Engine/EcsAlertFormatter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Bulwark.Core.Models;

namespace Bulwark.Core.Engine;

/// <summary>
/// 把一次安全事件 + 裁决格式化为 ECS(Elastic Common Schema)风格的结构化告警。
///
/// 目的:让磐垒可作为「传感器」接入 SIEM(Elastic / Splunk / OpenSearch 等)。
/// 输出为嵌套字典,序列化后即一条 ECS 兼容的 JSON 文档,字段遵循 ECS 命名约定:
///   @timestamp / event.* / process.* / process.code_signature.* / destination.* /
///   rule.* / threat.technique[] / threat.tactic[],并在 bulwark.* 下保留证据链等扩展。
///
/// 纯函数、无副作用、平台无关,便于单测。由服务端在事件处置后导出。
/// </summary>
public static class EcsAlertFormatter
{
    public const string EcsVersion = "8.11.0";

    /// <summary>构造一条 ECS 风格告警(嵌套字典)。</summary>
    public static Dictionary<string, object?> Format(SecurityEvent e, Verdict v)
    {
        var doc = new Dictionary<string, object?>
        {
            ["@timestamp"] = (e.TimestampUtc == default ? DateTime.UtcNow : e.TimestampUtc)
                .ToString("yyyy-MM-ddTHH:mm:ss.fffZ"),
            ["ecs"] = new Dictionary<string, object?> { ["version"] = EcsVersion },
            ["event"] = BuildEvent(e, v),
            ["message"] = BuildMessage(e, v),
            ["process"] = BuildProcess(e),
        };

        // 网络外联 -> destination
        if (e.Type == EventType.NetworkConnect && !string.IsNullOrEmpty(e.Target))
            doc["destination"] = BuildDestination(e.Target);

        // 命中规则 -> rule
        if (!string.IsNullOrEmpty(e.MatchedRuleNote))
            doc["rule"] = new Dictionary<string, object?> { ["name"] = e.MatchedRuleNote };

        // ATT&CK -> threat
        var threat = BuildThreat(e);
        if (threat is not null) doc["threat"] = threat;

        // 扩展命名空间:保留证据链与磐垒特有字段
        doc["bulwark"] = BuildBulwark(e, v);

        return doc;
    }

    private static Dictionary<string, object?> BuildEvent(SecurityEvent e, Verdict v)
    {
        string action = v.Action switch
        {
            VerdictAction.Block => "blocked",
            VerdictAction.Ask => "prompted",
            _ => "allowed"
        };
        // 阻止 = 拦截成功;其余视为放行/询问的处理成功
        var categories = new List<string>();
        if (e.HasThreatIndicator || v.Action == VerdictAction.Block)
            categories.Add("intrusion_detection");
        categories.Add(e.Type == EventType.NetworkConnect ? "network" : "process");

        return new Dictionary<string, object?>
        {
            ["kind"] = e.HasThreatIndicator || v.Action == VerdictAction.Block ? "alert" : "event",
            ["category"] = categories,
            ["action"] = action,
            ["type"] = new[] { EventTypeToEcs(e.Type) },
            ["outcome"] = "success",
            ["risk_score"] = e.RiskScore,
            ["severity"] = SeverityOf(e.RiskScore),
            ["provider"] = "Bulwark",
            ["module"] = e.Type.ToString(),
            ["reason"] = v.Source.ToString()
        };
    }

    private static Dictionary<string, object?> BuildProcess(SecurityEvent e)
    {
        var proc = new Dictionary<string, object?>
        {
            ["executable"] = e.ActorPath,
            ["name"] = SafeName(e.ActorPath),
            ["pid"] = e.ActorPid,
        };
        if (!string.IsNullOrEmpty(e.CommandLine)) proc["command_line"] = e.CommandLine;
        if (!string.IsNullOrEmpty(e.ActorHash))
            proc["hash"] = new Dictionary<string, object?> { ["sha256"] = e.ActorHash };

        proc["code_signature"] = new Dictionary<string, object?>
        {
            ["exists"] = e.ActorSigned || e.SignatureMismatch,
            ["trusted"] = e.ActorSigned && !e.SignatureMismatch && !e.CertRevoked && !e.SignedAfterCertExpiry,
            ["valid"] = e.ActorSigned && !e.SignatureMismatch,
            ["subject_name"] = e.ActorPublisher,
            ["status"] = SignatureStatus(e)
        };

        if (e.ParentPid != 0 || !string.IsNullOrEmpty(e.ParentPath))
        {
            proc["parent"] = new Dictionary<string, object?>
            {
                ["executable"] = e.ParentPath,
                ["name"] = SafeName(e.ParentPath),
                ["pid"] = e.ParentPid
            };
        }
        return proc;
    }

    private static Dictionary<string, object?> BuildDestination(string target)
    {
        var dest = new Dictionary<string, object?>();
        string host = target;
        int colon = target.LastIndexOf(':');
        if (colon > 0 && colon < target.Length - 1 && target.Skip(colon + 1).All(char.IsDigit))
        {
            host = target.Substring(0, colon);
            if (int.TryParse(target.Substring(colon + 1), out var port)) dest["port"] = port;
        }
        dest["address"] = host;
        if (System.Net.IPAddress.TryParse(host, out _)) dest["ip"] = host;
        else dest["domain"] = host;
        return dest;
    }

    private static Dictionary<string, object?>? BuildThreat(SecurityEvent e)
    {
        // 从证据链取去重的技战术(优先用已注解的 Technique 字段)
        var techIds = new List<string>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var ev in e.EvidenceChain)
            if (!string.IsNullOrEmpty(ev.Technique) && seen.Add(ev.Technique!))
                techIds.Add(ev.Technique!);

        if (techIds.Count == 0) return null;

        var techniques = new List<object>();
        var tactics = new List<object>();
        var tacticSeen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var id in techIds)
        {
            var t = AttackCatalog.Lookup(id);
            techniques.Add(new Dictionary<string, object?>
            {
                ["id"] = id,
                ["name"] = t?.Name ?? id
            });
            if (t is not null && tacticSeen.Add(t.Tactic))
                tactics.Add(new Dictionary<string, object?> { ["name"] = t.Tactic });
        }

        return new Dictionary<string, object?>
        {
            ["framework"] = "MITRE ATT&CK",
            ["technique"] = techniques,
            ["tactic"] = tactics
        };
    }

    private static Dictionary<string, object?> BuildBulwark(SecurityEvent e, Verdict v)
    {
        return new Dictionary<string, object?>
        {
            ["event_id"] = e.Id.ToString(),
            ["verdict"] = v.Action.ToString(),
            ["verdict_source"] = v.Source.ToString(),
            ["has_threat_indicator"] = e.HasThreatIndicator,
            ["user_mode_observed"] = e.UserModeObserved,
            ["reasons"] = e.RiskReasons.ToList(),
            ["techniques"] = e.Techniques.ToList(),
            ["evidence"] = e.EvidenceChain.Select(ev => new Dictionary<string, object?>
            {
                ["source"] = ev.Source,
                ["kind"] = ev.Kind.ToString(),
                ["description"] = ev.Description,
                ["score_delta"] = ev.ScoreDelta,
                ["technique"] = ev.Technique
            }).Cast<object>().ToList()
        };
    }

    private static string BuildMessage(SecurityEvent e, Verdict v)
    {
        string act = v.Action switch
        {
            VerdictAction.Block => "拦截",
            VerdictAction.Ask => "询问",
            _ => "放行"
        };
        return $"[{act}] {e.Type} {SafeName(e.ActorPath)}(pid={e.ActorPid}) -> {e.Target} (风险 {e.RiskScore})";
    }

    /// <summary>ECS event.severity:0-21 区间常用映射;这里按风险分粗分四级。</summary>
    private static int SeverityOf(int risk) => risk switch
    {
        >= 80 => 3, // critical/high
        >= 50 => 2, // medium
        > 0 => 1,   // low
        _ => 0
    };

    private static string SignatureStatus(SecurityEvent e)
    {
        if (!e.ActorSigned && !e.SignatureMismatch) return "unsigned";
        if (e.SignatureMismatch) return "tampered";
        if (e.CertRevoked) return "revoked";
        if (e.SignedAfterCertExpiry) return "expired";
        return "valid";
    }

    private static string EventTypeToEcs(EventType t) => t switch
    {
        EventType.ProcessCreate => "start",
        EventType.ProcessTerminate => "end",
        EventType.NetworkConnect => "connection",
        EventType.FileWrite => "change",
        EventType.FileDelete => "deletion",
        EventType.RegistryWrite => "change",
        EventType.ImageLoad => "info",
        EventType.RemoteThread => "access",
        _ => "info"
    };

    private static string SafeName(string? path)
    {
        if (string.IsNullOrEmpty(path)) return string.Empty;
        try { return System.IO.Path.GetFileName(path); }
        catch { return path; }
    }
}

File: cpp/shared/src/engine/EcsAlertFormatter.cpp
#include "bulwark/engine/EcsAlertFormatter.h"
#include "bulwark/engine/AttackCatalog.h"
#include "bulwark/json/JsonSupport.h"

#include <QDateTime>
#include <QFileInfo>
#include <QJsonArray>
#include <QSet>

namespace bulwark::engine {
namespace {

using bulwark::EventType;
using bulwark::SecurityEvent;
using bulwark::Verdict;
using bulwark::VerdictAction;

// File name of a path (empty for empty input); mirrors Path.GetFileName.
QString safeName(const QString& path) {
    if (path.isEmpty()) return QString();
    return QFileInfo(path).fileName();
}

// Lightweight IPAddress.TryParse stand-in (shared lib links Qt6::Core only, so
// no QHostAddress): IPv4 dotted-quad, or anything containing ':' as IPv6-ish.
bool isLikelyIp(const QString& host) {
    if (host.contains(QLatin1Char(':'))) return true;
    const QStringList parts = host.split(QLatin1Char('.'));
    if (parts.size() != 4) return false;
    for (const QString& p : parts) {
        bool ok = false;
        const int n = p.toInt(&ok);
        if (!ok || n < 0 || n > 255) return false;
    }
    return true;
}

// ECS event.severity: coarse 4-level mapping by risk score.
int severityOf(int risk) {
    if (risk >= 80) return 3;
    if (risk >= 50) return 2;
    if (risk > 0) return 1;
    return 0;
}

QString signatureStatus(const SecurityEvent& e) {
    if (!e.actorSigned && !e.signatureMismatch) return QStringLiteral("unsigned");
    if (e.signatureMismatch) return QStringLiteral("tampered");
    if (e.certRevoked) return QStringLiteral("revoked");
    if (e.signedAfterCertExpiry) return QStringLiteral("expired");
    return QStringLiteral("valid");
}

QString eventTypeToEcs(EventType t) {
    switch (t) {
        case EventType::ProcessCreate:    return QStringLiteral("start");
        case EventType::ProcessTerminate: return QStringLiteral("end");
        case EventType::NetworkConnect:   return QStringLiteral("connection");
        case EventType::FileWrite:        return QStringLiteral("change");
        case EventType::FileDelete:       return QStringLiteral("deletion");
        case EventType::RegistryWrite:    return QStringLiteral("change");
        case EventType::ImageLoad:        return QStringLiteral("info");
        case EventType::RemoteThread:     return QStringLiteral("access");
        case EventType::DnsQuery:         return QStringLiteral("protocol");
        default:                          return QStringLiteral("info");
    }
}

QJsonObject buildEvent(const SecurityEvent& e, const Verdict& v) {
    const QString action = v.action == VerdictAction::Block ? QStringLiteral("blocked")
                         : v.action == VerdictAction::Ask   ? QStringLiteral("prompted")
                                                            : QStringLiteral("allowed");
    QJsonArray categories;
    if (e.hasThreatIndicator || v.action == VerdictAction::Block)
        categories.append(QStringLiteral("intrusion_detection"));
    categories.append((e.type == EventType::NetworkConnect || e.type == EventType::DnsQuery)
                          ? QStringLiteral("network") : QStringLiteral("process"));

    QJsonArray typeArr;
    typeArr.append(eventTypeToEcs(e.type));

    QJsonObject ev;
    ev[QStringLiteral("kind")] =
        (e.hasThreatIndicator || v.action == VerdictAction::Block) ? QStringLiteral("alert")
                                                                   : QStringLiteral("event");
    ev[QStringLiteral("category")] = categories;
    ev[QStringLiteral("action")] = action;
    ev[QStringLiteral("type")] = typeArr;
    ev[QStringLiteral("outcome")] = QStringLiteral("success");
    ev[QStringLiteral("risk_score")] = e.riskScore;
    ev[QStringLiteral("severity")] = severityOf(e.riskScore);
    ev[QStringLiteral("provider")] = QStringLiteral("Bulwark");
    ev[QStringLiteral("module")] = bulwark::eventTypeToString(e.type);
    ev[QStringLiteral("reason")] = bulwark::verdictSourceToString(v.source);
    return ev;
}

QJsonObject buildProcess(const SecurityEvent& e) {
    QJsonObject proc;
    proc[QStringLiteral("executable")] = e.actorPath;
    proc[QStringLiteral("name")] = safeName(e.actorPath);
    proc[QStringLiteral("pid")] = e.actorPid;
    if (!e.commandLine.isEmpty()) proc[QStringLiteral("command_line")] = e.commandLine;
    if (!e.actorHash.isEmpty()) {
        QJsonObject h;
        h[QStringLiteral("sha256")] = e.actorHash;
        proc[QStringLiteral("hash")] = h;
    }
    QJsonObject sig;
    sig[QStringLiteral("exists")] = e.actorSigned || e.signatureMismatch;
    sig[QStringLiteral("trusted")] =
        e.actorSigned && !e.signatureMismatch && !e.certRevoked && !e.signedAfterCertExpiry;
    sig[QStringLiteral("valid")] = e.actorSigned && !e.signatureMismatch;
    sig[QStringLiteral("subject_name")] = e.actorPublisher;
    sig[QStringLiteral("status")] = signatureStatus(e);
    proc[QStringLiteral("code_signature")] = sig;

    if (e.parentPid != 0 || !e.parentPath.isEmpty()) {
        QJsonObject parent;
        parent[QStringLiteral("executable")] = e.parentPath;
        parent[QStringLiteral("name")] = safeName(e.parentPath);
        parent[QStringLiteral("pid")] = e.parentPid;
        proc[QStringLiteral("parent")] = parent;
    }
    return proc;
}

QJsonObject buildDestination(const QString& target) {
    QJsonObject dest;
    QString host = target;
    const int colon = target.lastIndexOf(QLatin1Char(':'));
    if (colon > 0 && colon < target.size() - 1) {
        const QString suffix = target.mid(colon + 1);
        bool allDigits = !suffix.isEmpty();
        for (const QChar& c : suffix)
            if (!c.isDigit()) { allDigits = false; break; }
        if (allDigits) {
            host = target.left(colon);
            bool ok = false;
            const int port = suffix.toInt(&ok);
            if (ok) dest[QStringLiteral("port")] = port;
        }
    }
    dest[QStringLiteral("address")] = host;
    if (isLikelyIp(host)) dest[QStringLiteral("ip")] = host;
    else dest[QStringLiteral("domain")] = host;
    return dest;
}

} // namespace
} // namespace bulwark::engine

namespace bulwark::engine {
namespace {

// Distinct annotated ATT&CK techniques from the evidence chain -> threat.*.
QJsonObject buildThreat(const SecurityEvent& e, bool& hasAny) {
    QStringList techIds;
    QSet<QString> seen;
    for (const auto& ev : e.evidenceChain) {
        if (ev.technique.isEmpty()) continue;
        const QString low = ev.technique.toLower();
        if (!seen.contains(low)) { seen.insert(low); techIds << ev.technique; }
    }
    hasAny = !techIds.isEmpty();
    if (!hasAny) return QJsonObject();

    QJsonArray techniques, tactics;
    QSet<QString> tacticSeen;
    for (const QString& id : techIds) {
        const auto t = AttackCatalog::lookup(id);
        QJsonObject tech;
        tech[QStringLiteral("id")] = id;
        tech[QStringLiteral("name")] = t.has_value() ? t->name : id;
        techniques.append(tech);
        if (t.has_value() && !tacticSeen.contains(t->tactic.toLower())) {
            tacticSeen.insert(t->tactic.toLower());
            QJsonObject tac;
            tac[QStringLiteral("name")] = t->tactic;
            tactics.append(tac);
        }
    }
    QJsonObject threat;
    threat[QStringLiteral("framework")] = QStringLiteral("MITRE ATT&CK");
    threat[QStringLiteral("technique")] = techniques;
    threat[QStringLiteral("tactic")] = tactics;
    return threat;
}

// bulwark.* extension namespace: verdict + full evidence chain.
QJsonObject buildBulwark(const SecurityEvent& e, const Verdict& v) {
    QJsonObject b;
    b[QStringLiteral("event_id")] = bulwark::json::guidToString(e.id);
    b[QStringLiteral("verdict")] = bulwark::verdictActionToString(v.action);
    b[QStringLiteral("verdict_source")] = bulwark::verdictSourceToString(v.source);
    b[QStringLiteral("has_threat_indicator")] = e.hasThreatIndicator;
    b[QStringLiteral("user_mode_observed")] = e.userModeObserved;
    b[QStringLiteral("reasons")] = QJsonArray::fromStringList(e.riskReasons);
    b[QStringLiteral("techniques")] = QJsonArray::fromStringList(e.techniques);
    QJsonArray evidence;
    for (const auto& ev : e.evidenceChain) {
        QJsonObject o;
        o[QStringLiteral("source")] = ev.source;
        o[QStringLiteral("kind")] = bulwark::evidenceKindToString(ev.kind);
        o[QStringLiteral("description")] = ev.description;
        o[QStringLiteral("score_delta")] = ev.scoreDelta;
        if (!ev.technique.isEmpty()) o[QStringLiteral("technique")] = ev.technique;
        evidence.append(o);
    }
    b[QStringLiteral("evidence")] = evidence;
    return b;
}

QString buildMessage(const SecurityEvent& e, const Verdict& v) {
    const QString act = v.action == VerdictAction::Block ? QString::fromUtf8("\xE6\x8B\xA6\xE6\x88\xAA")   // 拦截
                      : v.action == VerdictAction::Ask   ? QString::fromUtf8("\xE8\xAF\xA2\xE9\x97\xAE")   // 询问
                                                          : QString::fromUtf8("\xE6\x94\xBE\xE8\xA1\x8C"); // 放行
    return QString::fromUtf8("[%1] %2 %3(pid=%4) -> %5 (\xE9\xA3\x8E\xE9\x99\xA9 %6)") // 风险
        .arg(act, bulwark::eventTypeToString(e.type), safeName(e.actorPath))
        .arg(e.actorPid)
        .arg(e.target)
        .arg(e.riskScore);
}

} // namespace

QJsonObject EcsAlertFormatter::format(const bulwark::SecurityEvent& e, const bulwark::Verdict& v) {
    const QDateTime ts = e.timestampUtc.isValid() ? e.timestampUtc.toUTC()
                                                  : nowUtc();
    QJsonObject doc;
    doc[QStringLiteral("@timestamp")] =
        ts.toString(QStringLiteral("yyyy-MM-ddTHH:mm:ss.zzz")) + QStringLiteral("Z");
    QJsonObject ecs;
    ecs[QStringLiteral("version")] = QString::fromLatin1(EcsVersion);
    doc[QStringLiteral("ecs")] = ecs;
    doc[QStringLiteral("event")] = buildEvent(e, v);
    doc[QStringLiteral("message")] = buildMessage(e, v);
    doc[QStringLiteral("process")] = buildProcess(e);

    if (e.type == EventType::NetworkConnect && !e.target.isEmpty())
        doc[QStringLiteral("destination")] = buildDestination(e.target);

    if (e.type == EventType::DnsQuery && !e.target.isEmpty()) {
        QJsonObject question;
        question[QStringLiteral("name")] = e.target;
        QJsonObject dns;
        dns[QStringLiteral("type")] = QStringLiteral("query");
        dns[QStringLiteral("question")] = question;
        doc[QStringLiteral("dns")] = dns;
    }

    if (!e.matchedRuleNote.isEmpty()) {
        QJsonObject rule;
        rule[QStringLiteral("name")] = e.matchedRuleNote;
        doc[QStringLiteral("rule")] = rule;
    }

    bool hasThreat = false;
    const QJsonObject threat = buildThreat(e, hasThreat);
    if (hasThreat) doc[QStringLiteral("threat")] = threat;

    doc[QStringLiteral("bulwark")] = buildBulwark(e, v);
    return doc;
}

} // namespace bulwark::engine
